### ISSUE TITLE
copilot: add reasoning through reinforcement learning self-play (absolute-zero-reasoner method)

### ISSUE THREAD

**Issue by davidbasilefilho** [davidbasilefilho/atlas#6]
**Prompt Title:** Project: Implement Absolute-Zero-Reasoner for Atlas Model Optimization

**Role:** You are an expert AI programmer. Your task is to implement a reinforcement learning loop based on the "Absolute-Zero-Reasoner" paper to optimize the `Atlas` deep learning model.

**Context:**

  * **Target Project (`Atlas`):** A PyTorch implementation of a transformer model designed for long-term memory. The full project context, architecture, and usage instructions are detailed in the `/readme.md` file within the repository. You must read and understand it before beginning.
  * **Core Methodology (`Absolute-Zero-Reasoner`):** A method for improving model performance through reinforcement learning and self-play, guided by Monte Carlo Tree Search (MCTS). The reference paper is: `https://arxiv.org/pdf/2505.03335.pdf`. Your implementation should adapt its principles.

**Primary Objective:**
Create a self-play optimization loop that intelligently searches for the optimal `AtlasConfig` to maximize the model's performance on its evaluation suite. The "actions" in the self-play game will be modifications to the model's configuration parameters, guided by MCTS and structured reasoning.

-----

#### 1\. Create the Optimization Orchestrator (`optimize.py`)

This script is the entry point for the MCTS self-play loop.

  - **MCTS Implementation**: Implement a standard MCTS algorithm. Define a `Node` class and the `Selection`, `Expansion`, `Simulation`, and `Backpropagation` steps.
  - **State Representation**: A state is an instance of the `AtlasConfig` dataclass from `model.py`.
  - **Action Generation (Expansion Step)**: When expanding a node, you (the coding agent) will act as the "reasoner". You will be prompted with the current `AtlasConfig` and must generate new candidate configurations based on reasoning. Your reasoning must be in `<R>...</R>` tags, followed by a list of Python dictionaries representing the configuration modifications.

#### 2\. Implement the Simulation and Reward Function

  - **Simulation Step**: A simulation involves running a fast training and evaluation cycle for a given `AtlasConfig`.
  - **Reward Calculation**: Create a function to convert the evaluation metrics into a single scalar reward. This reward drives the MCTS search.

#### 3\. Data Collection for Reasoner Improvement

  - **Logging**: During self-play, log the data generated by the MCTS. For each completed search, save the initial state, the resulting search policy, and the final reward.

-----

#### `optimize.py` Structure

Your `optimize.py` script should contain the core MCTS logic.

  - **Node Class**: Define a class to represent each node in the search tree.
    ```python
    from model import AtlasConfig # Assuming AtlasConfig is importable

    class Node:
        def __init__(self, config: AtlasConfig, parent=None, action: dict = None):
            self.config = config          # The AtlasConfig for this state
            self.parent = parent
            self.action = action          # The modification that led to this node
            self.children = []
            self.visits = 0
            self.value = 0.0              # Cumulative reward from simulations
            self.is_fully_expanded = False
    ```
  - **MCTS Loop**: Structure your main loop around the select-expand-simulate-backpropagate cycle. Use an action-selection strategy like PUCT (Polynomial Upper Confidence Trees) to balance exploration and exploitation during the selection phase.

#### Inter-Process Communication and Evaluation

To ensure stability and encapsulation, run each simulation in a separate, isolated process.

1.  **Modify `main.py`**:
      - Adapt `main.py` to accept a path to a temporary JSON file that defines an `AtlasConfig`. For example: `python main.py --mode both --config_path /path/to/temp_config.json --debug`.
      - At the end of a full evaluation run, modify `main.py` to print a single JSON string to `stdout` containing the key metrics. This provides a machine-readable output.
        ```json
        {"perplexity": 10.5, "needle_accuracy": 0.9, "memory_capacity_score": 85.0}
        ```
2.  **`optimize.py` Simulation Step**:
      - Use Python's `subprocess` module to call `main.py`.
      - In your simulation function, write the node's `AtlasConfig` to a temporary JSON file.
      - Execute the `main.py` script as a subprocess and capture its `stdout`.
      - Parse the JSON output from `stdout` to get the evaluation metrics for the reward calculation.

#### Reward Calculation

Implement a clear, well-defined reward function within `optimize.py`.

```python
def calculate_reward(metrics: dict) -> float:
    """Calculates a scalar reward from evaluation metrics."""
    # Normalize perplexity (lower is better). Clip to avoid division by zero.
    # Assume max perplexity is ~100 for normalization purposes.
    ppl = metrics.get("perplexity", 100.0)
    ppl_score = max(0, 1 - (ppl / 100.0))

    # Needle accuracy is expected to be in [0, 1].
    needle_acc = metrics.get("needle_accuracy", 0.0)

    # Normalize memory capacity score (e.g., if it's out of 100).
    mem_score = metrics.get("memory_capacity_score", 0.0) / 100.0

    # Weighted sum for the final reward.
    reward = (0.4 * ppl_score) + (0.4 * needle_acc) + (0.2 * mem_score)
    return reward
```

#### Data Logging

For the `reasoner_training_data.jsonl` file, use a structured JSON Lines format. Each line should be a JSON object representing one MCTS search.

```json
{"initial_config": {"hidden_size": 512, "num_layers": 8, ...}, "mcts_policy": [{"action": {"polynomial_degree": 4}, "visits": 50}, {"action": {"hidden_size": 256, "num_layers": 6}, "visits": 25}], "best_reward": 0.82}
```

-----

### Deliverables

1.  A new Python script: **`optimize.py`** containing the MCTS optimization loop.
2.  Modifications to existing project files (**`main.py`**, **`evaluate.py`**) to allow for programmatic execution via configuration files and to output a final JSON summary.
3.  A `README.md` section explaining how to run `optimize.py` and detailing the architecture of the self-play system, including the precise reward function.
4.  A file **`reasoner_training_data.jsonl`** containing at least 10 examples of the collected `(state, policy, reward)` data from a sample run.
